### PR TITLE
As a Developer, I should be able to configure Administrator roles via fixtures

### LINKED ISSUE #54: Making the admin role configurable in Fixture
It would be helpful if you could specify the role an admin has in the fixtures of the admin.
```yml
admin_user:
          options:
            custom:
              - email: 'test@sylius.com'
                username: 'sylius'
                password: 'sylius'
                enabled: true
                locale_code: 'en_US'
                first_name: 'Testuser'
                last_name: 'Brille24'
                role: 'configurator'
```

### PR DESCRIPTION
**Pull request by Tomanhez**
fixes #54

### CODE REVIEW

File: src/Factory/AdminUserExampleFactory.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+final class AdminUserExampleFactory extends BaseAdminUserExampleFactory implements ExampleFactoryInterface
+{
+    /** @var RepositoryInterface */
+    private $administrationRoleRepository;
+
+    public function __construct(FactoryInterface $userFactory, RepositoryInterface $administrationRoleRepository, string $localeCode)
+    {
+        parent::__construct($userFactory, $localeCode);
```
Comment (Zales0123) [inline src/Factory/AdminUserExampleFactory.php]: Missing blank line below

File: src/Factory/AdminUserExampleFactory.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Sylius\RbacPlugin\Entity\AdministrationRoleInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+final class AdminUserExampleFactory extends BaseAdminUserExampleFactory implements ExampleFactoryInterface
+{
+    /** @var RepositoryInterface */
+    private $administrationRoleRepository;
+
+    public function __construct(FactoryInterface $userFactory, RepositoryInterface $administrationRoleRepository, string $localeCode)
```
Comment (Zales0123) [inline src/Factory/AdminUserExampleFactory.php]: Arguments list should be split into multiple lines

File: src/Fixture/AdminUserFixture.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+use Sylius\Bundle\CoreBundle\Fixture\AdminUserFixture as BaseAdminUserFixture;
+use Sylius\Bundle\FixturesBundle\Fixture\FixtureInterface;
+use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
+
+final class AdminUserFixture extends BaseAdminUserFixture implements FixtureInterface
+{
+    protected function configureResourceNode(ArrayNodeDefinition $resourceNode): void
+    {
+        parent::configureResourceNode($resourceNode);
```
Comment (Zales0123) [inline src/Fixture/AdminUserFixture.php]: Missing blank line below

File: tests/Behat/Context/Setup/AdminUserContext.php [hunk 4]
```diff
@@ -6,17 +6,34 @@
...
+    /** @var ExampleFactoryInterface */
+    private $adminUserExampleFactory;
+
+    /** @var UserRepositoryInterface */
+    private $adminUserRepository;
+
+    /** @var SharedStorageInterface */
+    private $sharedStorage;
+
+    public function __construct(ObjectManager $administratorManager, ExampleFactoryInterface $adminUserExampleFactory, UserRepositoryInterface $adminUserRepository, SharedStorageInterface $sharedStorage)
```
Comment (Zales0123) [inline tests/Behat/Context/Setup/AdminUserContext.php]: The same here, arguments list in multiple lines would be much more readable

File: features/administrators/adding_new_administrator_account_with_administration_role.feature [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is already an Administration role "No sections access" in the system
+
+    @ui
+    Scenario: Add new administrator account with administration role in the system
+        When I want to add a new administrator "good.father@abc.com" named "Good" last name "Father" and locale "en_US"
+        And this administrator has administration role "No sections access"
+        Then I should seeing administrator account "good.father@abc.com" named "Good" last name "Father" and locale "en_US"
```
Comment (Zales0123) [inline features/administrators/adding_new_administrator_account_with_administration_role.feature]: `Then administrator account "good.father@abc.com" should have first name "Good", last name "Father", and locale "en_US"`

File: features/administrators/adding_new_administrator_account_with_administration_role.feature [hunk 6]
```diff
@@ -0,0 +1,16 @@
+@managing_administration_roles
+Feature: Adding a new Administrator account with administration role
+    In order to add new Administrator account to the system
```
Comment (Zales0123) [inline features/administrators/adding_new_administrator_account_with_administration_role.feature]: `In order to manage permissions of new Administrators`

File: features/administrators/adding_new_administrator_account_with_administration_role.feature [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    As an Administrator
+    I want to be able to add new Administrator account
```
Comment (Zales0123) [inline features/administrators/adding_new_administrator_account_with_administration_role.feature]: `I want to be able to add a new Administrator account with administration role`

File: features/administrators/adding_new_administrator_account_with_administration_role.feature [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is already an Administration role "No sections access" in the system
+
+    @ui
+    Scenario: Add new administrator account with administration role in the system
+        When I want to add a new administrator "good.father@abc.com" named "Good" last name "Father" and locale "en_US"
+        And this administrator has administration role "No sections access"
+        Then I should seeing administrator account "good.father@abc.com" named "Good" last name "Father" and locale "en_US"
+        And I should seeing that this administrator has administration role "No sections access"
```
Comment (Zales0123) [inline features/administrators/adding_new_administrator_account_with_administration_role.feature]: `And this administrator should have administration role "No sections access"`

File: tests/Behat/Context/Setup/AdminUserContext.php [hunk 9]
```diff
@@ -30,4 +47,50 @@ public function thisAdministratorHasRole(
...
+        ]);
+
+        $this->adminUserRepository->add($adminUser);
+        $this->sharedStorage->set('administrator', $adminUser);
+    }
+
+    /**
+     * @Then I should seeing administrator account :email named :firstName last name :lastname and locale :locale
+     */
+    public function iShouldSeeAdministratorAccountNamedLastNameAndLocale(string $email, string $firstName, string $lastname, string $locale): void
```
Comment (Zales0123) [inline tests/Behat/Context/Setup/AdminUserContext.php]: This should not be placed in the `Setup` context - as well as the check should be done with UI

File: tests/Behat/Context/Setup/AdminUserContext.php [hunk 10]
```diff
@@ -30,4 +47,50 @@ public function thisAdministratorHasRole(
...
+            $user->getUsername() === $email &&
+            $user->getLocaleCode() === $locale &&
+            $user->isEnabled() === true
+        );
+    }
+
+    /**
+     * @Given /^I should seeing that (this administrator) has (administration role "[^"]+")$/
+     */
+    public function iShouldSeeThatThisAdministratorHasAdministrationRole(
```
Comment (Zales0123) [inline tests/Behat/Context/Setup/AdminUserContext.php]: As well as this one

File: tests/Behat/Context/Setup/AdminUserContext.php [hunk 11]
```diff
@@ -30,4 +47,50 @@ public function thisAdministratorHasRole(
...
+            $user->getFirstName() === $firstName &&
+            $user->getLastName() === $lastname &&
+            $user->getUsername() === $email &&
+            $user->getLocaleCode() === $locale &&
+            $user->isEnabled() === true
+        );
+    }
+
+    /**
+     * @Given /^I should seeing that (this administrator) has (administration role "[^"]+")$/
```
Comment (Zales0123) [inline tests/Behat/Context/Setup/AdminUserContext.php]: `@Given` -> `@Then`

File: features/administrators/adding_new_administrator_account_with_administration_role.feature (hunk 6 above)
Comment (GSadee) [reply features/administrators/adding_new_administrator_account_with_administration_role.feature]: **a**dministrators?

File: features/administrators/adding_new_administrator_account_with_administration_role.feature [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    As an Administrator
+    I want to be able to add new Administrator account
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is already an Administration role "No sections access" in the system
+
+    @ui
+    Scenario: Add new administrator account with administration role in the system
+        When I want to add a new administrator "good.father@abc.com" named "Good" last name "Father" and locale "en_US"
```
Comment (GSadee) [inline features/administrators/adding_new_administrator_account_with_administration_role.feature]: `When I want to add a new administrator "good.father@abc.com" named "Good" with "Father" last name and "en_US" locale` ?

File: tests/Behat/Context/Setup/AdminUserContext.php [hunk 13]
```diff
@@ -30,4 +47,50 @@ public function thisAdministratorHasRole(
 
         $this->administratorManager->flush();
     }
+
+    /**
+     * @When I want to add a new administrator :email named :firstName last name :lastname and locale :locale
+     */
+    public function iWantToAddAnNewAdministratorNamedLastName(string $email, string $firstName, string $lastname, string $locale): void
```
Comment (GSadee) [inline tests/Behat/Context/Setup/AdminUserContext.php]: `$locale` => `$localeCode`

File: src/Factory/AdminUserExampleFactory.php [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        parent::__construct($userFactory, $localeCode);
+
+        $this->administrationRoleRepository = $administrationRoleRepository;
+    }
+
+    protected function configureOptions(OptionsResolver $resolver): void
+    {
+        parent::configureOptions($resolver);
+
+        $resolver->setDefined('administration_role');
```
Comment (GSadee) [inline src/Factory/AdminUserExampleFactory.php]:
```
$resolver
    ->setAllowedTypes('administration_role', ['string', AdministrationRoleInterface::class, 'null'])
    ->setNormalizer('administration_role', LazyOption::findOneBy($this->administrationRoleRepository, 'name'))
;
```

File: features/administrators/adding_new_administrator_account_with_administration_role.feature [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+    I want to be able to add a new Administrator account with administration role
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is already an Administration role "No sections access" in the system
+
+    @ui
+    Scenario: Adding a new administrator account with administration role in the system
+        When I add a new administrator "good.father@abc.com" named "Good" with "Father" last name and "No sections access" administration role
+        Then administrator account "good.father@abc.com" should have first name "Good" and last name "Father"
```
Comment (GSadee) [inline features/administrators/adding_new_administrator_account_with_administration_role.feature]:
```suggestion
        Then the administrator account "good.father@abc.com" should have "Good" first name and "Father" last name
```

File: tests/Behat/Context/Setup/AdminUserContext.php [hunk 16]
```diff
@@ -20,7 +21,7 @@ public function __construct(ObjectManager $administratorManager)
     }
 
     /**
-     * @Given /^(this administrator) has (administration role "[^"]+")$/
+     * @When /^(this administrator) has (administration role "[^"]+")$/
```
Comment (GSadee) [inline tests/Behat/Context/Setup/AdminUserContext.php]: Unnecessary change

File: tests/Behat/Context/Setup/AdminUserContext.php [hunk 17]
```diff
@@ -30,4 +31,14 @@ public function thisAdministratorHasRole(
 
         $this->administratorManager->flush();
     }
+
+    /**
+     * @Then /^(this administrator) should have (administration role "([^"]*)")$/
```
Comment (GSadee) [inline tests/Behat/Context/Setup/AdminUserContext.php]: `Setup` context isn't a right place for `Then` step

File: tests/Behat/Context/UI/ManagingAdministratorsContext.php [hunk 18]
```diff
@@ -25,16 +31,54 @@ final class ManagingAdministratorsContext implements Context
...
         $this->indexPage = $indexPage;
         $this->updatePage = $updatePage;
         $this->adminUserRepository = $adminUserRepository;
+        $this->administratorManager = $administratorManager;
+        $this->adminUserExampleFactory = $adminUserExampleFactory;
+        $this->sharedStorage = $sharedStorage;
+    }
+
+    /**
+     * @When I add a new administrator :email named :firstName with :lastname last name and :administrationRole administration role
```
Comment (GSadee) [inline tests/Behat/Context/UI/ManagingAdministratorsContext.php]: It is not a Ui step, it is rather a domain one. You should probably move this step to other context in `Domain` namespace

File: features/administrators/adding_new_administrator_account_with_administration_role.feature [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+Feature: Adding a new Administrator account with administration role
+    In order to manage permissions of new Administrators
+    As an Administrator
+    I want to be able to add a new Administrator account with administration role
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is already an Administration role "No sections access" in the system
+
+    @ui
```
Comment (GSadee) [inline features/administrators/adding_new_administrator_account_with_administration_role.feature]: I guess that it should be a `@domain` scenario

File: src/Factory/AdminUserExampleFactory.php [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    public function create(array $options = []): AdminUserInterface
+    {
+        $user = parent::create($options);
+
+        if (!isset($options['administration_role'])) {
+            return $user;
+        }
+
+        /** @var AdministrationRoleInterface $administrationRole */
+        $administrationRole = $this->administrationRoleRepository->find($options['administration_role']);
```
Comment (GSadee) [inline src/Factory/AdminUserExampleFactory.php]: Is it still needed after adding `setNormalizer`?

File: tests/Behat/Context/UI/ManagingAdministratorsContext.php [hunk 21]
```diff
@@ -79,4 +123,20 @@ public function administratorShouldHaveNoRoleAssigned(string $email): void
     {
         Assert::same('', $this->indexPage->getAdministrationRoleOfAdminWithEmail($email));
     }
+
+    /**
+     * @Then administrator account :email should have first name :firstName and last name :lastname
+     */
+    public function administratorAccountShouldHaveFirstNameAndLastName(string $email, string $firstName, string $lastname): void
```
Comment (GSadee) [inline tests/Behat/Context/UI/ManagingAdministratorsContext.php]: It is not a `Ui` step also

File: tests/Behat/Resources/suites.yml [hunk 22]
```diff
@@ -43,6 +58,7 @@ default:
                 - Tests\Sylius\RbacPlugin\Behat\Context\Setup\AdminUserContext
 
                 - Tests\Sylius\RbacPlugin\Behat\Context\UI\ManagingAdministrationRolesContext
+                - Tests\Sylius\RbacPlugin\Behat\Context\UI\ManagingAdministratorsContext
```
Comment (GSadee) [inline tests/Behat/Resources/suites.yml]: Not needed I guess

**Comment by lchrusciel** [comment]
Thanks, Tomasz! :tada:
